src/arithmetic/algebraic_expression/algebraic_expression_eval.c

```diff
@@ -0,0 +1,255 @@
+		// Perform addition.
+		if(GrB_eWiseAdd_Matrix_Semiring(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, res, b,
+										GrB_NULL) != GrB_SUCCESS) {
+			printf("Failed adding operands, error:%s\n", GrB_error());
+			assert(false);
+		}
+
+		// Reset descriptor.
+		GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
+	}
+
+	if(inter != GrB_NULL) GrB_Matrix_free(&inter);
+	GrB_free(&desc);
+	return res;
+}
+
+static GrB_Matrix _Eval_Mul(const AlgebraicExpression *exp, GrB_Matrix res) {
+	assert(exp &&
+		   AlgebraicExpression_ChildCount(exp) > 1 &&
+		   AlgebraicExpression_OperationCount(exp, AL_EXP_MUL) == 1);
+
+	GrB_Matrix A;
+	GrB_Matrix B;
+	GrB_Info info;
+	GrB_Index nvals;
+	GrB_Descriptor desc;
+	AlgebraicExpression *left = CHILD_AT(exp, 0);
+	AlgebraicExpression *right = CHILD_AT(exp, 1);
+
+	GrB_Descriptor_new(&desc);  // Descriptor used for transposing operands.
+
+	if(left->type == AL_OPERATION) {
+		assert(left->operation.op = AL_EXP_TRANSPOSE);
+		GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+		left = CHILD_AT(left, 0);
+	}
+	A = left->operand.matrix;
+
+	if(right->type == AL_OPERATION) {
+		assert(right->operation.op = AL_EXP_TRANSPOSE);
+		GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+		right = CHILD_AT(right, 0);
+	}
+	B = right->operand.matrix;
+
+	if(B == IDENTITY_MATRIX) {
+		// B is the identity matrix, Perform A * I.
+		info = GrB_Matrix_apply(res, GrB_NULL, GrB_NULL, GrB_IDENTITY_BOOL, A, desc);
+		if(info != GrB_SUCCESS) {
+			// If the multiplication failed, print error info to stderr and exit.
+			fprintf(stderr, "Encountered an error in matrix multiplication:\n%s\n", GrB_error());
+			assert(false);
+		}
+	} else {
+		// Perform multiplication.
+		info = GrB_mxm(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, A, B, desc);
+		if(info != GrB_SUCCESS) {
+			// If the multiplication failed, print error info to stderr and exit.
+			fprintf(stderr, "Encountered an error in matrix multiplication:\n%s\n", GrB_error());
+			assert(false);
+		}
+	}
+
+	GrB_Matrix_nvals(&nvals, res);
+
+	// Reset descriptor.
+	GrB_Descriptor_set(desc, GrB_INP0, GxB_DEFAULT);
+	GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
+
+	uint child_count = AlgebraicExpression_ChildCount(exp);
+	for(uint i = 2; i < child_count; i++) {
+		right = CHILD_AT(exp, i);
+		if(right->type == AL_OPERATION) {
+			assert(right->operation.op = AL_EXP_TRANSPOSE);
+			GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+			right = CHILD_AT(right, 0);
+		}
+		B = right->operand.matrix;
+
+		if(B != IDENTITY_MATRIX) {
+			// Perform multiplication.
+			info = GrB_mxm(res, GrB_NULL, GrB_NULL, GxB_LOR_LAND_BOOL, res, B, desc);
+			if(info != GrB_SUCCESS) {
+				// If the multiplication failed, print error info to stderr and exit.
+				fprintf(stderr, "Encountered an error in matrix multiplication:\n%s\n", GrB_error());
+				assert(false);
+			}
+		}
+		GrB_Matrix_nvals(&nvals, res);
+		if(nvals == 0) break;
+
+		// Reset descriptor.
+		GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
```

Comment:
Nit, but we could reset at the start of the loop and remove the call at 182.

Reply:
OK